packages/yew-router/src/router.rs

```diff
@@ -30,78 +31,66 @@ impl PartialEq for RouterState {
     }
 }
 
-#[doc(hidden)]
-pub enum Msg {
-    ReRender,
+pub(crate) enum RouterStateAction {
+    Navigate,
+    ReplaceHistory(AnyHistory),
 }
 
-/// The Router component.
-///
-/// This provides [`History`] context to its children and switches.
-///
-/// You only need one `<Router />` for each application.
-pub struct Router {
-    _listener: HistoryListener,
-    history: AnyHistory,
-    ctr: u32,
-}
-
-impl Component for Router {
-    type Message = Msg;
-    type Properties = RouterProps;
-
-    fn create(ctx: &Context<Self>) -> Self {
-        let link = ctx.link().clone();
-
-        let listener = ctx
-            .props()
-            .history
-            .listen(move || link.send_message(Msg::ReRender));
+impl Reducible for RouterState {
+    type Action = RouterStateAction;
 
-        Self {
-            _listener: listener,
-            history: ctx.props().history.clone(),
-            ctr: 0,
-        }
-    }
+    fn reduce(self: Rc<Self>, action: Self::Action) -> Rc<Self> {
+        match action {
+            RouterStateAction::Navigate => Self {
+                history: self.history(),
+                ctr: self.ctr + 1,
+            }
+            .into(),
 
-    fn update(&mut self, _ctx: &Context<Self>, msg: Self::Message) -> bool {
-        match msg {
-            Msg::ReRender => {
-                self.ctr += 1;
-                true
+            RouterStateAction::ReplaceHistory(history) => Self {
+                history,
+                ctr: self.ctr + 1,
             }
+            .into(),
         }
     }
+}
 
-    fn changed(&mut self, ctx: &Context<Self>) -> bool {
-        let link = ctx.link().clone();
-
-        if self.history != ctx.props().history {
-            self._listener = ctx
-                .props()
-                .history
-                .listen(move || link.send_message(Msg::ReRender));
-
-            self.history = ctx.props().history.clone();
-
-            true
-        } else {
-            false
-        }
+/// The Router component.
+///
+/// This provides [`History`] context to its children and switches.
+///
+/// You only need one `<Router />` for each application.
+#[function_component(Router)]
+pub fn router(props: &RouterProps) -> Html {
+    let state = use_reducer(|| RouterState {
```

Comment:
```suggestion
    let RouterProps { history, children } = props.clone();
    let state = use_reducer(|| RouterState {
```

Reply:
I think you could just de-structured props in the first line of the component and clone them since all of them seem like they are getting cloned